### PR TITLE
script: Add features and rustdoc checks

### PR DESCRIPTION
**Pull request by febo**
### Problem

Currently there is no automated way to check if every combination of features is working properly on rust crates (interface or client) nor if the code on `rustdoc` compiles.

### Solution

This PR adds two script to perform additional steps:
* `check-features`: uses `cargo-hack` to perform a feature powerset check.
* `rustdoc`: runs `rustdoc` build the documentation.

These scripts run by default on CI and can also be run locally.

Thanks @kevinheavey for suggesting adding these!

### CODE REVIEW

File: interface/Cargo.toml [hunk 1]
```diff
@@ -47,3 +47,4 @@ frozen-abi = [
     "dep:solana-logger",
     "solana-pubkey/frozen-abi"
 ]
+test = ["solana-pubkey/std"]
```
Comment (kevinheavey) [inline interface/Cargo.toml:50]: Consider adding `solana-system-interface = { path = ".", features = "test" }` to dev-dependencies.

If this was just added for `cargo hack` (I don't see it used anywhere else), you could instead do the following:
- remove this feature
- add `solana-pubkey = { version = "2.1", features = ["std"] }` to dev-dependencies
- add `--all-targets`  to the `cargo hack` command

Should probably add `--all-targets` either way

File: .github/actions/setup/action.yml [hunk 2]
```diff
@@ -70,6 +70,10 @@ runs:
         version: ${{ env.SOLANA_VERSION }}
         cache: true
 
+    - name: Install 'cargo-hack'
```
Comment (lorisleiva) [inline .github/actions/setup/action.yml:73]: Should we add an if statement here to ensure we only install this if `toolchain` is set to `lint`?

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -40,6 +40,12 @@ jobs:
       - name: Lint Client Rust
         run: pnpm clients:rust:lint
 
+      - name: Check Client Rust Features
```
Comment (lorisleiva) [inline .github/workflows/main.yml:43]: I think this is fine but just curious why we can't just stick these checks inside the lint script?

Comment (febo) [reply .github/workflows/main.yml:43]: We could, I was doing that first. The only caveat is that if we want to pass specific command-line arguments, it won't be easy. Should I add both steps to the `lint` script?

File: .github/actions/setup/action.yml (hunk 2 above)
Comment (lorisleiva) [reply .github/actions/setup/action.yml:73]: Whichever toolchain they need. But otherwise JS CI steps will end up installing this when they don't need to.

File: .github/workflows/main.yml (hunk 3 above)
Comment (lorisleiva) [reply .github/workflows/main.yml:43]: I'm not sure how often these scripts are used in isolation. From an outsider perspective, I kinda just want a script or two that tells me if my code is alright plus one that test my code. That's why you added the new "check" script. Because the amount of scripts is becoming overwhelming. I think "categorising" these within the broader concept of "linting" the code makes sense for me but again, I'm not an experienced Rustacian so I'm not sure this is what people will expect.

Comment (febo) [reply .github/workflows/main.yml:43]: Yeah, exactly, that is the reason that I added a `check` that runs all of them since that is probably what you want locally. I left them separate for the CI so when it fails, it is more obvious which one has a problem.

Comment (febo) [reply .github/workflows/main.yml:43]: But I am not sure, I think we need @joncinque and @buffalojoec opinion as well. ☺️

Comment (joncinque) [reply .github/workflows/main.yml:43]: There are quite a few scripts, with a lot of copy-pasta between them. It might be nice to only have language-specific scripts, ie. `scripts/rust` and `scripts/js` that can be reused for all rust packages, or all js packages, etc.

Personally, I've found it slightly annoying to update _all_ the scripts when making a change, but maybe it doesn't happen too often.

**Review by joncinque — COMMENTED** [review]
Looks good to me (with Loris's comments addressed)

File: package.json [hunk 4]
```diff
@@ -12,12 +12,18 @@
     "clients:js:lint": "zx ./scripts/client/lint-js.mjs",
     "clients:js:publish": "zx ./scripts/client/publish-js.mjs",
     "clients:js:test": "zx ./scripts/client/test-js.mjs",
+    "clients:rust:check": "pnpm clients:rust:format && pnpm clients:rust:lint && pnpm clients:rust:features && pnpm clients:rust:rustdoc",
```
Comment (buffalojoec) [inline package.json:15]: I'm not against an aggregate `pnpm` script that runs format, lint, and test all in one go.

File: scripts/client/rustdoc.mjs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+const args = cliArguments();
+const manifestPath = path.join(
+  workingDirectory,
+  'clients',
+  'rust',
+  'Cargo.toml'
+);
+
+await $`RUSTDOCFLAGS="--cfg docsrs -D warnings" cargo doc --manifest-path ${manifestPath} --all-features --no-deps ${args}`;
```
Comment (buffalojoec) [inline scripts/client/rustdoc.mjs:13]: I could be missing some context here, but `cargo test` (_and_ `test-sbf`) will test-compile your rust docs. It's just `clippy` that doesn't. `fmt` will also make sure your docs/comments are formatted according to the standard. So I think we're pretty covered on rustdocs.

File: .github/workflows/main.yml (hunk 3 above)
Comment (buffalojoec) [reply .github/workflows/main.yml:43]: I mostly agree on aggregate language-specific scripts, but I think it's probably easier to do one script per language, per library (ie. `scripts/program`, `scripts/client/rust`, `scripts/client/js`). Mostly because the program will use `-sbf` related commands and configs. Not a huge deal either way, though.

I don't love the fact that the number of scripts is growing, but it does make sense to me for `clippy` and `check` (`hack`) to be separate jobs, invokable separately. Including it in the `format_and_lint` CI job is totally fine, though.

File: scripts/client/rustdoc.mjs (hunk 5 above)
Comment (kevinheavey) [reply scripts/client/rustdoc.mjs:13]: `cargo test` and `cargo fmt` don't cover everything that cargo doc covers, such as bad links in doc comments

Comment (buffalojoec) [reply scripts/client/rustdoc.mjs:13]: Seems gratuitous to me, but if you guys want the extra coverage I won't object.

Comment (kevinheavey) [reply scripts/client/rustdoc.mjs:13]: It tends to catch errors quite often

File: .github/actions/setup/action.yml (hunk 2 above)
Comment (febo) [reply .github/actions/setup/action.yml:73]: Switch to use `lint` in all scripts related to linting.

File: .github/workflows/main.yml (hunk 3 above)
Comment (febo) [reply .github/workflows/main.yml]: Yeah, having many similar scripts is not ideal. I had a chat with Loris and we came up with the following. Divide the scripts by language, as Jon suggested (`script/js` and `scripts/rust`), and have a "configuration" script specific for each package when needed.

The scripts directory would look like:
```
scripts
   |
   +- js
   |  |
   |  +- format.mjs
   |  +- lint.mjs
   |  +- publish.mjs
   |  +- test.mjs
   |
   +- rust
   |  |
   |  +- build.mjs
   |  +- clean.mjs
   |  +- dump.mjs
   |  +- format.mjs
   |  +- lint-clippy.mjs
   |  +- lint-docs.mjs
   |  +- lint-features.mjs
   |  +- lint.mjs
   |  +- test.mjs
   |  
   +- crate
      |
      +- client.mjs
      +- interface.mjs
      +- program.mjs
```

All rust scripts then accept a `"crate"` parameter, which specify the custom "configuration" script to use (folder `crate`). In these scripts we can add any specific variables for the crate.

In summary, we will have generic language scripts + specific configuration ones. Since this is a bigger refactoring, the idea is to do this in a separate PR.

File: scripts/client/lint-rust.mjs
Comment (buffalojoec) [inline scripts/client/lint-rust.mjs:1]: nit: `lint-rust-all.mjs`?

**Review by buffalojoec — COMMENTED** [review]
Wfm, as long as we're planning to move toward grouping the scripts by language, as mentioned in discussions. Lots of scripts!

File: scripts/client/lint-rust.mjs
Comment (lorisleiva) [reply scripts/client/lint-rust.mjs:1]: I have a slight preference over `lint` than `lint-all` because it makes it clearer that `lint` is a parent of `lint-*`. Not massively fussed though.

Comment (buffalojoec) [reply scripts/client/lint-rust.mjs:1]: I also don't feel strongly, but the `-all` for me is usually indicative that it includes multiple combined steps.

Comment (lorisleiva) [reply scripts/client/lint-rust.mjs:1]: 🤷‍♂️😅

Comment (febo) [reply scripts/client/lint-rust.mjs:1]: `lint` keeps it "symmetric" with the JS ones. In any case, `lint-rust` will go away when we restructure the scripts, so perhaps we can remove the `lint` that call all of them and add all three under the same script on the `package.json` instead.

File: scripts/client/lint-rust.mjs [hunk 6]
```diff
@@ -1,28 +1,19 @@
...
-  workingDirectory,
-} from '../utils.mjs';
+// Script to lint the Rust client crate.
+//
+// This script runs the following sub-scripts:
+// - lint-rust-clippy.mjs
+// - lint-rust-features.mjs
+// - lint-rust-docs.mjs
+//
+// Command line arguments are only passed to clippy.
```
[the remark this answers is not in the document]
Comment (febo) [reply scripts/client/lint-rust.mjs:8]: I like this, keeps it simple and avoid "hardcoded" behaviour.

**Review by buffalojoec — APPROVED** [review]
Ship it!

File: scripts/client/lint-rust-clippy.mjs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  'rust',
+  'Cargo.toml'
+);
+
+// Check the client using Clippy.
+if (fix) {
+  await $`cargo ${toolchain} clippy --manifest-path ${manifestPath} --fix ${lintArgs}`;
+} else {
+  await $`cargo ${toolchain} clippy --manifest-path ${manifestPath} ${lintArgs}`;
+}
```
Comment (joncinque) [inline scripts/client/lint-rust-clippy.mjs:28]: nit: the other script does this in one line using a ternary operator. I'm only pointing this out as another piece of evidence that we shouldn't duplicate the scripts :smiling_imp:

**Review by joncinque — COMMENTED** [review]
Just a couple of little nits, but this should be ready!

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -32,13 +32,19 @@ jobs:
       - name: Setup Environment
         uses: ./.github/actions/setup
         with:
-          toolchain: format, lint
+          toolchain: format, lint, test
```
[the remark this answers is not in the document]
Comment (febo) [reply .github/workflows/main.yml:35]: Good catch, it is not used anymore.

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -50,13 +56,19 @@ jobs:
...
```
[the remark this answers is not in the document]
Comment (febo) [reply .github/workflows/main.yml:59]: Same here, not used anymore.

File: scripts/client/lint-rust-clippy.mjs (hunk 7 above)
Comment (febo) [reply scripts/client/lint-rust-clippy.mjs:28]: Very good point. 😅

**Review by joncinque — APPROVED** [review]
